app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,17 +1,212 @@
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
 
 /**
  * Utility class for encrypting submissions during the SaveToDiskTask.
  *
  * @author mitchellsundt@gmail.com
  */
+
 public class EncryptionUtils {
 
+    private static final int PASSPHRASE_LENGTH = 32;
+    private static final String SECRET_NAME = "secret";
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String ALGORITHM = KeyProperties.KEY_ALGORITHM_AES;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String BLOCK_MODE = KeyProperties.BLOCK_MODE_CBC;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String PADDING = KeyProperties.ENCRYPTION_PADDING_PKCS7;
+    private static KeyStore keystore = null;
+
+    //Gets the SecretKey from the Android KeyStore (creates a new one the first time)
+    private static SecretKey getKey()
+            throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException,
+            UnrecoverableEntryException, InvalidAlgorithmParameterException {
+        if(keystore == null) {
+            keystore = KeyStore.getInstance("AndroidKeyStore");
+            keystore.load(null);
+        }
+
+        KeyStore.Entry existingKey = keystore.getEntry(SECRET_NAME, null);
+        if (existingKey instanceof KeyStore.SecretKeyEntry) {
+            return ((KeyStore.SecretKeyEntry) existingKey).getSecretKey();
+        }
+
+        //Create key
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+            KeyGenParameterSpec.Builder builder = new KeyGenParameterSpec.Builder(SECRET_NAME,
+                    KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT);
+            builder.setBlockModes(BLOCK_MODE);
+            builder.setEncryptionPaddings(PADDING);
+            builder.setUserAuthenticationRequired(false);
+            builder.setRandomizedEncryptionRequired(true);
+
+            KeyGenerator generator = KeyGenerator.getInstance(ALGORITHM);
+            generator.init(builder.build());
+            return generator.generateKey();
+        }
+
+        return null;
+    }
+
+    //Generate a random passphrase
+    public static byte[] generatePassphrase() {
+        Random random;
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
+            try {
+                random = SecureRandom.getInstanceStrong();
+            } catch (NoSuchAlgorithmException e) {
+                random = new Random();
+            }
+        } else {
+            random = new Random();
+        }
+
+        byte[] result = new byte[PASSPHRASE_LENGTH];
+
+        while (true) {
+            random.nextBytes(result);
+
+            //Make sure there are no zeroes in the passphrase (bad for SQLCipher)
```

Comment:
would be great to add a reference on why it's bad for SqlCipher.

Reply:
No problem